### PR TITLE
Make Broadcast.result_style work on styles with fields.

### LINKED ISSUE #50937: Broadcast.result_style(x::S, y::S) does not work unless S() is a constructor
[This line](https://github.com/JuliaLang/julia/blame/90b4eedbf4b2beae0b12e6b8317e15ef0bb91126/base/broadcast.jl#L460) does not work with broadcast styles that have fields, eg

```julia
import Broadcast: BroadcastStyle

struct MyStyle <: BroadcastStyle
    some_info
end

BroadcastStyle(x::MyStyle, y::MyStyle) = MyStyle(compute_field(x.some_info, y.some_info))
```

A possible fix (in user code) is defining
```julia
Broadcast.result_style(::MyStyle, ::MyStyle)
```
but that function does not seem to be part of the API.

A possible fix (in Julia) would be 
```julia
result_join(::Any, ::Any, x::S, y::S) where S<:BroadcastStyle = x # just pick the first
```

I wanted to ask if this matches the intention of #20740 before making a PR.

### PR DESCRIPTION
**Pull request by tpapp**
Fixes #50937.

### CODE REVIEW

**Review by vtjnash — COMMENTED** [review]
Can you fix line 444 too (the result_style constructor), which seems to have the same issue to my cursory reading? Actually, maybe this should call `result_style(s1, s2)` here, and then that would be the function generally overloaded to merge two distinct values?

File: test/broadcast.jl [hunk 1]
```diff
@@ -1142,6 +1142,19 @@ end
...
+    i::Int
+end
+# asymmetry intended
+Base.BroadcastStyle(a::MyBroadcastStyleWithField, b::MyBroadcastStyleWithField) = a.i
+
+@testset "issue #50937: styles that have fields" begin
+    @test Broadcast.result_style(MyBroadcastStyleWithField(1), MyBroadcastStyleWithField(1)) ==
+        MyBroadcastStyleWithField(1)
+    @test_throws ErrorException Broadcast.result_style(MyBroadcastStyleWithField(1),
+                                                       MyBroadcastStyleWithField(2)) # order matters
```
Comment (vtjnash) [inline test/broadcast.jl:1155]:
```suggestion
                                                       MyBroadcastStyleWithField(2))
```

**Comment by tpapp** [comment]
Thanks for pointing it out, I fixed that too, and the test had an error (should have subtyped the `struct`), that is now also fixed.

**Comment by vtjnash** [comment]
@nanosoldier `runbenchmarks("broadcast", vs=":master")`

**Pull request merged**